File: py_scripts/py_scripts/prompt_extract_user_message.py
from pathlib import Path
from typing import Union


def prompt_extract_user_message(file_path: Union[str, Path]) -> str:
    """
    Extract the USER MESSAGE section from a Markdown file.

    Reads a Markdown file and extracts all text starting from the
    '# USER MESSAGE' heading through the end of the file.
    The extracted text is returned as a single string with original
    line breaks preserved.

    Parameters
    ----------
    file_path : str or pathlib.Path
        Path to the markdown (.md) file.

    Returns
    -------
    str
        The USER MESSAGE text.

    Raises
    ------
    ValueError
        If no '# USER MESSAGE' section is found.
    """
    lines = Path(file_path).read_text(encoding="utf-8").splitlines()

    # Locate the USER MESSAGE heading (exact match, like R's `lines == ...`)
    try:
        start_idx = lines.index("# USER MESSAGE")
    except ValueError:
        raise ValueError("No '# USER MESSAGE' section found in the file. Is there a typo?")

    # Move to the first line of content under the heading
    start_idx += 1

    # Extract everything to end of file
    return "\n".join(lines[start_idx:])

File: r/prompt_extract_user_message.R
#' Extract the USER MESSAGE section from a markdown file
#'
#' Reads a markdown file and extracts all text starting from the
#' `# User Message` heading through the end of the file.
#' The extracted text is returned as a single character string with
#' original line breaks preserved.
#'
#' @param file_path Character string. Path to the markdown (.md) file.
#'
#' @return A character string containing the USER MESSAGE text.
#'
#' @details
#' This function treats the markdown file as plain text and relies on
#' heading detection rather than a markdown parser. It assumes:
#' \itemize{
#'   \item The USER MESSAGE section is introduced with `# User Message`
#'   \item All relevant content appears after this heading
#' }
#'
#' If the USER MESSAGE heading is not found, the function errors.
#'
#' @examples
#' \dontrun{
#' user_message <- prompt_extract_user_message(
#'   "Template Prompt - Lesson Revisions.md"
#' )
#'
#' cat(user_message)
#' }
#'
#' @export
prompt_extract_user_message <- function(file_path) {
  
  # Read the markdown file line-by-line
  lines <- readLines(file_path, encoding = "UTF-8")
  
  # Locate the USER MESSAGE heading
  start_idx <- which(lines == "# USER MESSAGE")
  
  # Error if the heading is not found
  if (length(start_idx) == 0) {
    stop("No '# USER MESSAGE' section found in the file. Is there a typo?")
  }
  
  # Move to the first line of content under the heading
  start_idx <- start_idx + 1
  
  # Extract everything from USER MESSAGE to end of file
  user_message <- paste(
    lines[start_idx:length(lines)],
    collapse = "\n"
  )
  
  return(user_message)
}
